File: Encapsulation/Encapsulation/Communication/DataModel/VectorClock.cs
using Collector.Communication.DataModel;
using NLog;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace Encapsulation.Communication.DataModel
{
    internal class VectorClock
    {
        private Logger m_ApplicationLogger;
        public ConcurrentDictionary<string, int> Clock { get; private set; } = new ConcurrentDictionary<string, int>();
        public string ProcessKey { get; private set; }

        public VectorClock(Logger applicationLogger, string currentProcessKey) 
        {
            m_ApplicationLogger = applicationLogger;
            ProcessKey = currentProcessKey;
            Clock.AddOrUpdate(currentProcessKey, 0, (k, v) => 0);
        } 

        public VectorClock(Logger applicationLogger, string currentProcessKey, VectorClock other) : this(applicationLogger, currentProcessKey) 
        {
            AddRange(other);
        }

        public void AddRange(VectorClock clock)
        {
            if (clock == null)
                throw new ArgumentNullException(nameof(clock));
            foreach (var element in clock.Clock)
                Clock.AddOrUpdate(element.Key, element.Value, (k, v) => element.Value);
        }

        public void Increase()
        {
            Clock.AddOrUpdate(ProcessKey, 0, (k, v) => v + 1);
        }

        public bool IsNewer(VectorClock other)
        {
            var key = other.ProcessKey;

            if (!Clock.ContainsKey(key))
            {
                return true;
            }
            else
            {
                return Clock[key] < other.Clock[key];
            }
        }

        public bool IsNewer(string key, List<ProcessState> states)
        {
            var singleState = states.FirstOrDefault(x => x.Key.Equals(key));

            if (singleState != null)
            {
                var currentValue = Clock.GetValueOrDefault(key);
                return currentValue < singleState.Value;
            }
            return false;
        }

        public List<ProcessState> ToProcessState()
        {
            var processStates = new List<ProcessState>();
            foreach (var entry in Clock.Keys)
            {
                // #SucheFreunde...
                var processState = new ProcessState();
                // #LadeFreunde...
                processState.Key = entry;
                processState.Value = Clock[entry];
                processStates.Add(processState);
                // #KeinenGefunden---Sorry...
            }
            return processStates;
        }

        public void AddSingleProcessState(string key, List<ProcessState> states)
        {
            var singleState = states.FirstOrDefault(x => x.Key.Equals(key));
            var currentValue = Clock.GetValueOrDefault(key);

            if (currentValue == null || currentValue < singleState.Value)
            {
                Clock.AddOrUpdate(key, singleState.Value, (k, v) => { return currentValue; });
            }
        }

        public void AddProcessStates(List<ProcessState> processStates)
        {
            foreach (var entry in processStates)
            {
                var key = entry.Key;
                var value = entry.Value;

                Clock.AddOrUpdate(key, value, (k, v) => {
                    return v > value ? v : value;
                });
            }
        }

        public void AddVectorClock(VectorClock other)
        {
            foreach (var key in other.Clock.Keys)
            {
                var value = other.Clock[key];

                Clock.AddOrUpdate(key, value, (k, v) => {
                    return v > value ? v : value;
                });
            }
        }
    }
}

File: QoRManager/Shared/src/Helper/VectorClock.java
package Helper;

import Network.DataModel.CommunicationMessages;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class VectorClock {
    private final Map<String, Integer> _Clock;
    private final String _ProcessKey;

    public VectorClock(String currentProcessKey) {
        this(currentProcessKey, new HashMap<>());
    }

    public VectorClock(String currentProcessKey, VectorClock other) {
        this(currentProcessKey, other.getClock());
    }

    public VectorClock(String currentProcessKey, Map<String, Integer> clock) {
        _ProcessKey = currentProcessKey;
        _Clock = new ConcurrentHashMap<>(clock);
        _Clock.putIfAbsent(currentProcessKey, 1);
    }

    public Map<String, Integer> getClock() {
        return new HashMap<>(_Clock);
    }

    public int getValueForProcessID(String id) {
        return _Clock.getOrDefault(id, -1);
    }

    public String getProcessKey(){
        return _ProcessKey;
    }

    public void addRange(VectorClock clock) {
        if (clock == null)
            throw new NullPointerException("The given vector clock is null.");
        for(var key : clock.getClock().keySet()) {
            _Clock.putIfAbsent(key, clock.getClock().get(key));
        }
    }

    public synchronized void increase() {
        _Clock.computeIfPresent(_ProcessKey, (k, v) -> v + 1);
    }

    public boolean isNewer(VectorClock other) {
        var key = other.getProcessKey();

        if (!_Clock.containsKey(key)) {
            return true;
        }
        else {
            return _Clock.get(key) < other.getClock().get(key);
        }
    }

    public List<CommunicationMessages.ProcessState> toProcessState() {
        synchronized (_Clock) {
            var processStates = new ArrayList<CommunicationMessages.ProcessState>();
            for(var key : _Clock.keySet())
            {
                // #SucheFreunde...
                var processState = CommunicationMessages.ProcessState.newBuilder()
                        // #LadeFreunde...
                        .setKey(key)
                        .setValue(_Clock.get(key))
                        .build();
                processStates.add(processState);
                // #KeinenGefunden---Sorry...
            }
            return processStates;
        }
    }

    public void addSingleProcessState(VectorClock other) {
        var key = other.getProcessKey();
        var otherProcessState = other.getClock().get(key);

        if(!_Clock.containsKey(key) || otherProcessState > _Clock.get(key)) {
            _Clock.put(key, otherProcessState);
        }
    }


    public void addProcessStates(List<CommunicationMessages.ProcessState> processStates) {
        for(var entry : processStates) {
            var key = entry.getKey();
            var value = entry.getValue();

            if (_Clock.containsKey(key)) {
                var oldValue = _Clock.get(key);
                if (oldValue > value) {
                    continue;
                }
                _Clock.put(key, value);
            }
            else {
                _Clock.put(key, value);
            }
        }
    }

    public void addVectorClock(VectorClock other) {
        var otherClock = other.getClock();
        for(var key : otherClock.keySet()) {
            var value = otherClock.get(key);

            if (_Clock.containsKey(key)) {
                var oldValue = _Clock.get(key);
                if (oldValue > value) {
                    continue;
                }
                _Clock.put(key, value);
            }
            else {
                _Clock.put(key, value);
            }
        }
    }
}
